idf_build_apps/app.py

```diff
@@ -449,29 +386,70 @@ def build(
         depends_on_components=None,  # type: list[str] | str | None
         check_component_dependencies=False,  # type: bool
     ):  # type: (...) -> bool
-        cmake_vars = self.build_prepare()
+        LOGGER.debug('=> Preparing...')
+        if self.work_dir != self.app_dir:
+            if os.path.exists(self.work_dir):
+                LOGGER.debug('==> Work directory %s exists, removing', self.work_dir)
+                if not self.dry_run:
+                    shutil.rmtree(self.work_dir)
+            LOGGER.debug('==> Copying app from %s to %s', self.app_dir, self.work_dir)
+            if not self.dry_run:
+                shutil.copytree(self.app_dir, self.work_dir)
+
+        if os.path.exists(self.build_path):
+            LOGGER.debug('==> Build directory %s exists, removing', self.build_path)
+            if not self.dry_run:
+                shutil.rmtree(self.build_path)
+
+        if not self.dry_run:
+            os.makedirs(self.build_path)
+
+        sdkconfig_file = os.path.join(self.work_dir, 'sdkconfig')
+        if os.path.exists(sdkconfig_file):
+            LOGGER.debug('==> Removing sdkconfig file: %s', sdkconfig_file)
+            if not self.dry_run:
+                os.unlink(sdkconfig_file)
 
         if self.build_log_path:
             LOGGER.info('=> Writing build log to %s', self.build_log_path)
+
+        if self.dry_run:
+            LOGGER.debug('==> Skipping... (dry run)')
+            return True
+
+        if self.build_log_path:
             log_file = open(self.build_log_path, 'w')
         else:
             # delete manually later, used for tracking debugging info
             log_file = tempfile.NamedTemporaryFile('w', delete=False)
 
+        # additional env variables
+        # IDF_TARGET to bypass the idf.py build check
+        # SDKCONFIG_DEFAULTS to bypass the file not exists issue
+        additional_env_dict = {
+            'IDF_TARGET': self.target,
+            'SDKCONFIG_DEFAULTS': ';'.join(self.sdkconfig_defaults_list) if self.sdkconfig_defaults_list else ';',
+        }
+
+        # check if this app depends on components according to the project_description.json 'build_component' field.
+        # the file is generated by `idf.py reconfigure`.
+        common_args = [
+            sys.executable,
+            str(IDF_PY),
+            '-B',
+            self.build_path,
+            '-C',
+            self.work_dir,
+        ]
```

Comment:
I don't like this solution because the `IDF_TARGET` and `SDKCONFIG_DEFAULT` variables are not visible outside of idf_build_apps. Could we do 
```suggestion
        additional_env_dict = {
            'IDF_TARGET': self.target,
            #'SDKCONFIG_DEFAULTS': ';'.join(self.sdkconfig_defaults_list) if self.sdkconfig_defaults_list else ';',
        }

        # check if this app depends on components according to the project_description.json 'build_component' field.
        # the file is generated by `idf.py reconfigure`.
        sdkconfig_var = self.sdkconfig_path if self.sdkconfig_path else ';'
        common_args = [
            sys.executable,
            str(IDF_PY),
            '-D',
            'SDKCONFIG_DEFAULTS=' + sdkconfig_var,
            '-B',
            self.build_path,
            '-C',
            self.work_dir,
        ]
```
I'm not sure whether this is a good solution, though

Reply:
Thanks, your idea makes more sense. I'll apply your suggestions.